src/nvim/ex_docmd.c

```diff
@@ -9952,3 +9953,104 @@ bool cmd_can_preview(char_u *cmd)
 
   return false;
 }
+
+/// Return the list of the commands in the specified buffer,
+/// if no buffer is specified, then return list of global(non-buffer) the
+/// commands
+///
+/// @param[in] buf the specified buffer, if NULL then global commands required
+///
+/// @return The array of the commands and the information of each command
+ArrayOf(Dictionary) commands_array(buf_T *buf)
+{
+  Array rv = ARRAY_DICT_INIT;
+  int i = 0, j = 0;
+  char Arg[2] = "";
+  garray_T *gap;
+  if (buf == NULL) {
+    gap = &ucmds;
+  } else {
+    gap = &buf->b_ucmds;
+  }
+  for (i = 0; i < gap->ga_len; i++) {
+    Dictionary dic = ARRAY_DICT_INIT;
+    bool ck = false, ck2 = false;
+    char Range[10] = "";
+    ucmd_T *cmd = USER_CMD_GA(gap, i);
+
+    // Name
+    PUT(dic, "name", STRING_OBJ(cstr_to_string((const char *)cmd->uc_name)));
+
+    // Argument
+    switch (cmd->uc_argt & (EXTRA|NOSPC|NEEDARG)) {
+    case 0:                    Arg[0] = '0'; break;
+    case(EXTRA):               Arg[0] = '*'; break;
+    case(EXTRA|NOSPC):         Arg[0] = '?'; break;
+    case(EXTRA|NEEDARG):       Arg[0] = '+'; break;
+    case(EXTRA|NOSPC|NEEDARG): Arg[0] = '1'; break;
+    }
+    PUT(dic, "argument", STRING_OBJ(cstr_to_string((const char *)Arg)));
+
+    // Definition
+    PUT(dic, "definition",
+        STRING_OBJ(cstr_to_string((const char *)cmd->uc_rep)));
+
+    // Complete
+    for (j = 0; command_complete[j].expand != 0; j++) {
+      if (cmd->uc_compl == command_complete[j].expand) {
```

Comment:
This operation can be made more fast, command_complete is just a static array, we manually inserted the items in it, if the items there are added in sorted order then by using binary search O(log(n)) complexity can be achieved instead of O(n), so possibliity can be just arange the items in array, replace linear search with bin search

Reply:
I think the proper refactor would be to define the complete strings next to the constants, as a dense array of string constants (with NULL if some constant lacks a string). Probably, move `EXPAND_COMMANDS` etc enum from vim.c and `command_complete` array both to `ex_cmds_defs.h`. Compare to `typedef enum hlf_T` and  `hlf_names` in `highlight_defs.h`. Maybe an accessor function should be used, to do bounds checking (negative index should return NULL string).

Reply:
How about this possibility, for array command_complete, I add {INT_MIN, NULL} at the end, this element can be used to serve the end

Reply:
You were asking for buffer?, means if negetive buffer number is sent?

Reply:
Didn't get it

Reply:
I think I get it,you were asking to move the enums and command_complete array to ex_cmds_defs.h, is it, then define a function which is used to access the array, but it requires a lot of changes mainly in previous made functions, will it

Reply:
I mean, instead of having an array with enum-string pairs `{EXPAND_AUGROUP, "augroup"}` we should have an array _indexed_ by the enum, i e `command_complete[EXPAND_AUGROUP] == "augroup"`. But the values of this enum can be negative, it needs to be checked before accessing the array.

If it seems daunting, just leave the code as it is. This is not a hot code path anyway...

Reply:
Ohh, you mean hashing like, I can do it, but it may be easy if the api provides the hashing, or I leave it for now.

Reply:
there is hashtab.h, which I think provides functionality of hashing, but, where to initialize the hashtable its static, so I need an entry point to init hashtable

Reply:
Like if I was using c++, then I need something like, `map<int, string> hash;`

Reply:
No hashtab is needed, just make it a dense array like `hlf_names`. If names are missing, use NULL as an indicator.

Reply:
No, leave it here, because, there is a line, [her]([https://github.com/neovim/neovim/blob/611351677dba450fc1a312061572c44c7e3d6482/src/nvim/ex_docmd.c#L5866), which suggests this implementation being better

Reply:
and also there is a function, `char_u *get_user_cmd_complete(expand_T *xp, int idx)` which is used and returns the index, so present implementation is good, or we have to create a new structure by keeping this array too

Reply:
> No hashtab is needed, just make it a dense array like hlf_names. If names are missing, use NULL as an indicator.
I see it

Reply:
I didn't know such a thing is possible in C

Reply:
Yeah there is a thing exists, I didn't know of it

Reply:
hmm `get_user_cmd_complete` is bit tricky, but unless the order in `vim.h` has any significance, we could just reorder them so that named `EXPAND_` kinds are first.

Reply:
But the function get_user_cmd_complete poses difficulty, it returns complete according to index, so, I now think, even rearrangement values not possible now

Reply:
it is used in a struct, it has an arg for function pointer, so some element of that struct calls this function

Reply:
It is used by `ExpandGeneric` which sorts the results anyway, so reordering them wouldn't be an issue.

Reply:
yes, ExpandGeneric uses it, I also found it, so true, reordering is best way here

Reply:
I reorder them according to the value of the EXPAND_

Reply:
I sorted the list of commands according to the values of EXPAND_, now I want to test if it will be correct, will `get_user_cmd_complete` function pose no difficulty, where is its test

Reply:
Sorting the lines in reverse sorted is better, so then we have not to care for the 0 as the last element which acts as comparator for end, and also useful in binary search